File: interviewbit-arrays-first-missing-integer.py
class Solution:
    # @param A : list of integers
    # @return an integer
    def firstMissingPositive(self, A):
        A = [num if num >= 0 and num <= len(A) else 0 for num in A]
        i = 0
        while i < len(A) - 1:
            if A[i] != 0 and A[i] != i + 1:
                if A[i] == A[A[i] - 1]:
                    i += 1
                else:
                    temp = A[i]
                    A[i] = A[temp - 1]
                    A[temp - 1] = temp
            else:
                i += 1

        for i, num in enumerate(A):
            if num != i + 1:
                return i + 1

        return len(A) + 1

File: interviewbit-arrays-first-missing-integer.java
public class Solution {
    public int firstMissingPositive(int[] A) {
        int temp;
        for (int i = 0; i < A.length; i++) {
            // if a number is out of place, is in the range 1 to the length of the array, and is not
            // a duplicate of the number already in its correct place, swap them
            if (A[i] != i + 1 && A[i] > 0 && A[i] <= A.length && A[i] != A[A[i] - 1]) {
                temp = A[A[i] - 1];
                A[A[i] - 1] = A[i];
                A[i] = temp;
                // and stay in place to check the number that was just swapped into that position, too
                i--;
            }
        }
        for (int i = 0; i < A.length; i++) {
            if (A[i] != i + 1) {
                return i + 1;
            }
        }
        return A.length + 1;
    }
}
